src/Tools/dotnet-monitor/CollectionRules/ActionOptionsDependencyAnalyzer.cs

```diff
@@ -0,0 +1,269 @@
+                }
+                Properties.Clear();
+            }
+        }
+
+        private sealed class ActionDependency
+        {
+            public ActionDependency(CollectionRuleActionOptions action, string resultName)
+            {
+                Action = action;
+                ResultName = resultName;
+            }
+
+            public CollectionRuleActionOptions Action { get; }
+
+            public string ResultName { get; }
+
+            public string GetActionResultToken() => string.Concat(ActionReferencePrefix,
+                Action.Name, Separator, ResultName, SubstitutionSuffix);
+        }
+
+        public ActionOptionsDependencyAnalyzer(CollectionRuleContext context)
+        {
+            _ruleContext = context ?? throw new ArgumentNullException(nameof(context));
+        }
+
+        public IList<CollectionRuleActionOptions> GetActionDependencies(int actionIndex)
+        {
+            EnsureDependencies();
+
+            if (_dependencies.TryGetValue(actionIndex, out Dictionary<string, PropertyDependency> properties))
+            {
+                return properties
+                    .SelectMany(p => p.Value.ActionDependencies)
+                    .DistinctBy(a => a.Action.Name, StringComparer.OrdinalIgnoreCase)
+                    .Select(a => a.Action).ToArray();
+            }
+            return Array.Empty<CollectionRuleActionOptions>();
+        }
+
+        public IDisposable SubstituteOptionValues(int actionIndex, object settings)
+        {
+            EnsureDependencies();
+
+            var revert = new PropertiesRevert(settings);
+
+            if (!_dependencies.TryGetValue(actionIndex, out Dictionary<string, PropertyDependency> properties))
+            {
+                return revert;
+            }
+
+            foreach (KeyValuePair<string, PropertyDependency> property in properties)
+            {
+                string newValue = property.Value.OriginalValue;
+                foreach(ActionDependency actionDependency in property.Value.ActionDependencies)
+                {
+                    string actionToken = actionDependency.GetActionResultToken();
+                    if (!_ruleContext.ActionResults.TryGetValue(actionDependency.Action.Name, out CollectionRuleActionResult results))
+                    {
+                        _ruleContext.Logger.InvalidResultReference(actionToken);
+                        continue;
+                    }
+                    if (!results.OutputValues.TryGetValue(actionDependency.ResultName, out string result))
+                    {
+                        _ruleContext.Logger.InvalidResultReference(actionToken);
+                        continue;
+                    }
+                    newValue = newValue.Replace(actionToken, result, StringComparison.OrdinalIgnoreCase);
+                }
+                property.Value.Property.SetValue(settings, newValue);
```

Comment:
I'm almost certain that you should not do this. The options are cached as part of a CollectionRuleOptions instance in the IOptionsCache<CollectionRuleOtions> service. If the same rule is applied to more than one process at the same time (very likely) then there will be multi-threading issues when setting and resetting these property values.

It's probably better (maybe even easier) to make a shallow clone of the options instance, change the values on the new instance, and return that new instance to be passed into the action.